rust/src/mask/object/serialization.rs

```diff
@@ -0,0 +1,188 @@
+
+        let config = MaskConfig::from_bytes(&self.config()).context("invalid MaskObject buffer")?;
+        let bytes_per_digit = config.bytes_per_digit();
+        let (data_length, overflows) = (self.digits() as usize).overflowing_mul(bytes_per_digit);
+        if overflows {
+            return Err(anyhow!(
+                "invalid MaskObject buffer: invalid mask config or digits field"
+            ));
+        }
+        let total_expected_length = DIGITS_FIELD.end + data_length;
+        if len < total_expected_length {
+            return Err(anyhow!(
+                "invalid buffer length: expected {} bytes but buffer has only {} bytes",
+                total_expected_length,
+                len
+            ));
+        }
+        Ok(())
+    }
+
+    pub fn len(&self) -> usize {
+        let config = MaskConfig::from_bytes(&self.config()).unwrap();
+        let bytes_per_digit = config.bytes_per_digit();
+        let data_length = self.digits() as usize * bytes_per_digit;
+        DIGITS_FIELD.end + data_length
+    }
+
+    pub fn digits(&self) -> u32 {
+        // UNWRAP SAFE: the slice is exactly 4 bytes long
+        u32::from_be_bytes(self.inner.as_ref()[DIGITS_FIELD].try_into().unwrap())
+    }
+
+    pub fn config(&self) -> &[u8] {
+        &self.inner.as_ref()[MASK_CONFIG_FIELD]
+    }
+
+    pub fn data(&self) -> &[u8] {
+        &self.inner.as_ref()[DIGITS_FIELD.end..self.len()]
+    }
+}
+
+impl<T: AsRef<[u8]> + AsMut<[u8]>> MaskObjectBuffer<T> {
+    pub fn set_digits(&mut self, value: u32) {
+        self.inner.as_mut()[DIGITS_FIELD].copy_from_slice(&value.to_be_bytes());
+    }
+    pub fn config_mut(&mut self) -> &mut [u8] {
+        &mut self.inner.as_mut()[MASK_CONFIG_FIELD]
+    }
+
+    pub fn data_mut(&mut self) -> &mut [u8] {
+        let end = self.len();
+        &mut self.inner.as_mut()[DIGITS_FIELD.end..end]
+    }
+}
+
+impl ToBytes for MaskObject {
+    fn buffer_length(&self) -> usize {
+        DIGITS_FIELD.end + self.config.bytes_per_digit() * self.data.len()
+    }
+
+    fn to_bytes<T: AsMut<[u8]>>(&self, buffer: &mut T) {
+        let mut writer = MaskObjectBuffer::new_unchecked(buffer.as_mut());
+        self.config.to_bytes(&mut writer.config_mut());
+        writer.set_digits(self.data.len() as u32);
+
+        let mut data = writer.data_mut();
+        let bytes_per_digit = self.config.bytes_per_digit();
+
+        for int in self.data.iter() {
+            // FIXME: this allocates a vec which is sub-optimal. See
+            // https://github.com/rust-num/num-bigint/issues/152
+            let bytes = int.to_bytes_le();
+            // This may panic if the data is invalid and contains
+            // integers that are bigger than what is expected by the
+            // configuration.
+            data[..bytes.len()].copy_from_slice(&bytes[..]);
+            // padding
+            for b in data.iter_mut().take(bytes_per_digit).skip(bytes.len()) {
+                *b = 0;
+            }
```

Comment:
the chance that padding is needed is `1/2**8 + 1/2**16 + ... + 1/2**(8*bytes_per_digit) = (1 - 2**(-8*bytes_per_digit)) / 255`, so roughly 0.4%, because the `BigUint`s are uniformly distributed. that means we don't add much overhead in the serialized format.

Reply:
> in PB-742 you mentioned going for a "not using BigUInt" approach, so what lead to changing your mind?